src/main/java/com/beecoders/ras/service/OrderService.java

```diff
@@ -88,22 +95,52 @@ public OrderDetailInfo getOrderDetailInfoById(Long id) {
 
     @Transactional
     public String payOrder(PayOrder payOrder) {
-        if (payOrder.getCardInfo() == null && payOrder.getPaymentType().equals(DEBIT.name()))
-            throw new IllegalPaymentException("Card should be present for DEBIT type payment");
+        if (Objects.isNull(payOrder.getCardInfo()) && payOrder.getPaymentType().equals(DEBIT.name()))
+            throw new IllegalPaymentException(NO_CARD_DEBIT_TYPE_ERROR_MESSAGE);
 
         Long orderId = payOrder.getOrderId();
         PaymentMethod paymentMethod = paymentMethodRepository.findByName(payOrder.getPaymentType())
                 .orElseThrow();
-        Order order = orderRepository.findById(orderId).orElseThrow(() ->
-                new EntityNotFoundException(String.format("Order with id (%s) not found", orderId)));
+        Order order = findOrderById(orderId);
 
         if (order.getFinishedAt() != null)
-            throw new IllegalPaymentException(String.format("Order with id (%s) already paid", orderId));
+            throw new IllegalPaymentException(String.format(ORDER_ALREADY_PAID_ERROR_MESSAGE, orderId));
 
         order.setPaymentMethod(paymentMethod);
         order.setFinishedAt(Timestamp.valueOf(LocalDateTime.now()));
 
         return PaymentType.valueOf(payOrder.getPaymentType()).getSuccessfulMessage();
     }
 
+    @Transactional
+    public void addPromocode(AddPromocode request) {
+        Order order = findOrderById(request.getOrderId());
+
+        if (!Objects.isNull(order.getPromocode()))
+            throw new IllegalUsingPromocodeException(ALREADY_USED_PROMOCODE_ERROR_MESSAGE);
+
+        Promocode promocode = promocodeRepository.findByCode(request.getPromocode())
+                .orElseThrow(() -> new EntityNotFoundException(String.format(PROMOCODE_NOT_FOUND_ERROR_MESSAGE,
+                        request.getPromocode())));
+
+        if (promocode.getToDate().isBefore(LocalDate.now()))
+            throw new IllegalUsingPromocodeException(EXPIRED_PROMOCODE_ERROR_MESSAGE);
+
```

Comment:
You should also check `from_date` parameter of promocode to consider that `now` is in promocode date range